src/basefold_verifier/query_phase.rs

```diff
@@ -338,10 +377,124 @@ pub(crate) fn batch_verifier_query_phase<C: Config>(
+            builder.dyn_array(round.openings.len());
+        let low_values_buffer: Array<C, Array<C, Felt<C::F>>> =
+            builder.dyn_array(round.openings.len());
+        let high_values_buffer: Array<C, Array<C, Felt<C::F>>> =
+            builder.dyn_array(round.openings.len());
+        let log2_heights = builder.dyn_array(round.openings.len());
+        let minus_alpha_offsets = builder.dyn_array(round.openings.len());
+        let all_zero_slices = builder.dyn_array(round.openings.len());
+        let opening_heights = builder.dyn_array(round.openings.len());
+        let dimensions = builder.dyn_array(round.openings.len());
+
+        iter_zip!(
+            builder,
+            opened_values_buffer,
+            log2_heights,
+            round.openings,
+            low_values_buffer,
+            high_values_buffer,
+            minus_alpha_offsets,
+            all_zero_slices,
+            opening_heights,
+            dimensions,
+        )
+        .for_each(|ptr_vec, builder| {
+            let opening = builder.iter_ptr_get(&round.openings, ptr_vec[2]);
+            let log2_height: Var<C::N> =
+                builder.eval(opening.num_var + Usize::from(get_rate_log() - 1));
+            builder.iter_ptr_set(&log2_heights, ptr_vec[1], log2_height);
+            let width = opening.point_and_evals.evals.len();
+
+            let opened_value_len: Var<C::N> = builder.eval(width.clone() * two);
+            let opened_value_buffer = builder.dyn_array(opened_value_len);
+            builder.iter_ptr_set(
+                &opened_values_buffer,
+                ptr_vec[0],
+                opened_value_buffer.clone(),
+            );
+            let low_values = opened_value_buffer.slice(builder, 0, width.clone());
+            let high_values =
+                opened_value_buffer.slice(builder, width.clone(), opened_value_buffer.len());
+            builder.iter_ptr_set(&low_values_buffer, ptr_vec[3], low_values.clone());
+            builder.iter_ptr_set(&high_values_buffer, ptr_vec[4], high_values.clone());
+
+            let alpha_offset = builder.get(&input.batch_coeffs, batch_coeffs_offset.clone());
+            // Will need to negate the values of low and high
+            // because `fri_single_reduced_opening_eval` is
+            // computing \sum_i alpha^i (0 - opened_value[i]).
+            // We want \sum_i alpha^(i + offset) opened_value[i]
+            // Let's negate it here.
+            builder.assign(&alpha_offset, -alpha_offset);
+            builder.iter_ptr_set(&minus_alpha_offsets, ptr_vec[5], alpha_offset);
+            builder.assign(&batch_coeffs_offset, batch_coeffs_offset + width.clone());
+
+            let all_zero_slice = all_zeros.slice(builder, 0, width);
+            builder.iter_ptr_set(&all_zero_slices, ptr_vec[6], all_zero_slice);
+
+            let num_var = opening.num_var;
+            let height: Var<C::N> = builder.eval(num_var + Usize::from(get_rate_log() - 1));
```

Comment:
It seems to me that `opening_heights` is same as `log2_heights`. I think we should only keep one.